### PR TITLE
Keyboard is visible for a brief moment on request/send money page after getting back from currency search page

### PR DESCRIPTION
**Pull request by narefyev91**
### Details
Implement new [event handler](https://reactnavigation.org/docs/navigation-events/#beforeremove) for navigation which will close keyboard before user will go out from the screen

### Fixed Issues

$ https://github.com/Expensify/App/issues/15085
$ https://github.com/Expensify/App/issues/15085#issuecomment-1431562751

### Tests
1. Open the app in android
2. Open any report
3. Click on plus
4. Click on Request money/ send money/ split bill
5. Click on currency
6. On currency search page, when keyboard is displayed for search, click on 'X' to close and get back to previous page
7. **Note for QA:** On mWeb, this behaviour might still be slugging, its same as in production

- [ ] Verify that no errors appear in the JS console

### Offline tests
This feature doesn't change or is impacted by offline mode.

### QA Steps
Same as above.

- [ ] Verify that no errors appear in the JS console

### PR Author Checklist
- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android / native
    - [x] Android / Chrome
    - [x] iOS / native
    - [x] iOS / Safari
    - [x] MacOS / Chrome / Safari
    - [x] MacOS / Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is correct English and approved by marketing by adding the `Waiting for Copy` label for a copy review on the original GH to get the correct copy.
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/StyleUtils.js) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(themeColors.componentBG`)
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] I have checked off every checkbox in the PR author checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Web</summary>

https://user-images.githubusercontent.com/28352309/219352082-0408bf18-7fb7-43fc-a38a-9acb7a6ecc2a.mov

</details>

<details>
<summary>Mobile Web - Chrome</summary>

https://user-images.githubusercontent.com/28352309/219352812-e13c579f-4011-4094-a2c6-63002e7c5470.mp4

</details>

<details>
<summary>Mobile Web - Safari</summary>

https://user-images.githubusercontent.com/28352309/219352235-9c17596b-9dde-4985-a5fc-3e1f9adf2354.mov

</details>

<details>
<summary>Desktop</summary>

https://user-images.githubusercontent.com/28352309/219352128-94140253-efa0-43e1-bffa-9f22b147148b.mov

</details>

<details>
<summary>iOS</summary>

https://user-images.githubusercontent.com/28352309/219352204-20b597ed-426d-432d-9c1b-d8a4504e6996.mov

</details>

<details>
<summary>Android</summary>

https://user-images.githubusercontent.com/28352309/219353057-12fd5099-9fd2-479f-843b-c2ed52641e2d.mp4

</details>

### CODE REVIEW

**Review by mountiny — COMMENTED** [review]
This is looking good to me! I have noticed there is a slight delay for the numpad to go down on iOS Safari not sure if that is on main as well.

@0xmiroslav there was also an alternative proposal to add this listener to the ScreenWrapper so this handles all the pages. From the linked issue:
> We can solve issue in 2 ways - globally and locally.
For global one - we can add subscription to event beforeremove inside ScreenWrapper class and add in compose withKeyboardState HOC - which will guaranty that keyboard will be closed on any screen which was wrapped in ScreenWrapper
For local one - doing the same steps - but only under IOUCurrencySelection class

Curious what do you think of it @0xmiroslav? I am just worried we might not be able to catch some issues this might cause on some other pages, but it might be better to move this level up, what do you think?

**Comment by narefyev91** [comment]
> This is looking good to me! I have noticed there is a slight delay for the numpad to go down on iOS Safari not sure if that is on main as well.
> 
> @0xmiroslav there was also an alternative proposal to add this listener to the ScreenWrapper so this handles all the pages. From the linked issue:
> 
> > We can solve issue in 2 ways - globally and locally.
> > For global one - we can add subscription to event beforeremove inside ScreenWrapper class and add in compose withKeyboardState HOC - which will guaranty that keyboard will be closed on any screen which was wrapped in ScreenWrapper
> > For local one - doing the same steps - but only under IOUCurrencySelection class
> 
> Curious what do you think of it @0xmiroslav? I am just worried we might not be able to catch some issues this might cause on some other pages, but it might be better to move this level up, what do you think?

Yup for safari on mobile it's the same on main page. Happened because keyboard is coming from device, and when it's closing, web receive this info with delay and re-build UI. As a workaround it can be done fixes using setTimeout - but that's may cause some issues on slow devices

**Comment by mountiny** [comment]
Thanks for the context 🙇 I think we dont have to worry about this then!

File: src/components/ScreenWrapper/index.js [hunk 1]
```diff
@@ -45,6 +46,10 @@ class ScreenWrapper extends React.Component {
             this.props.onTransitionEnd();
             Navigation.setIsNavigating(false);
         });
```
Comment (mountiny) [inline src/components/ScreenWrapper/index.js:48]:
```suggestion
        });
        
```

**Review by mountiny — COMMENTED** [review]
@0xmiroslav I have asked @narefyev91 to move the logic up but if you have any concerns, feel free to raise them.

Would be great to test some other screens which this might influence it

**Comment by mountiny** [comment]
@0xmiroslav thank you very much!

> If we encounter the case of new screen which doesn't need keyboard dismiss while closing, we may introduce new prop like shouldDismissKeyboardBeforeClose in ScreenWrapper with default value to true.

I think in near future we could have form flow where there could be multiple pages in row which will also have autofocus in it, so I think we might want to get ready for that. 

Then the props would be useful I imagine.

@narefyev91 Could you try to implement the `shouldDismissKeyboardBeforeClose` props @0xmiroslav mentioned?

**Comment by mountiny** [comment]
@narefyev91 there is also merge conflict

**Comment by narefyev91** [comment]
@mountiny @0xmiroslav Yup i get you guys - btw we could not really worried about navigation between screen - navigation event handler will unmount keyboard before moving to next screen and in case based on your paradigm and controlling directly focus event in didMount such issues will not be happened
As an example same screen with currency and adding money - we have autofocus on mount - after choose currency we go back - keyboard dismissed, and input focused

https://user-images.githubusercontent.com/28352309/220642087-3e5eb2b6-46f0-4945-b6c0-5d1995321011.mov

**Comment by narefyev91** [comment]
> > navigation event handler will unmount keyboard before moving to next screen
> 
> yes, I had this concern below but I confirmed that keyboard dismisses instantly when navigate to new screen, unlike when navigate to previous screen which is already mounted:
> 
> limit
> 
> Actually, this is android issue.
> 
> On iOS, keyboard dismisses instantly always - either navigating to previous or next

yup - that's totally makes sense - @0xmiroslav @mountiny added additional prop to handle event handlers - let me know - if that works!

**Comment by mountiny** [comment]
Thanks everyone! 

@0xmiroslav will you be able to go through the PR review checklist?

File: src/components/ScreenWrapper/index.js [hunk 2]
```diff
@@ -55,6 +56,15 @@ class ScreenWrapper extends React.Component {
             this.setState({didScreenTransitionEnd: true});
             this.props.onEntryTransitionEnd();
         });
+
+        if (this.props.shouldDismissKeyboardBeforeClose) {
```
Comment (mountiny) [inline src/components/ScreenWrapper/index.js:60]: A comment explaining why we doing this might be helpful I would say

**Review by mountiny — DISMISSED** [review]
Codewise looks good to me!

File: src/components/ScreenWrapper/index.js (hunk 2 above)
Comment (narefyev91) [reply src/components/ScreenWrapper/index.js:60]: @mountiny we already have comment in prop types for this field - do we want to add more description?

**Comment by narefyev91** [comment]
@mountiny @0xmiroslav pushed description for prop

File: src/components/ScreenWrapper/index.js (hunk 2 above)
Comment (mountiny) [reply src/components/ScreenWrapper/index.js:63]: Agreed with Miroslav, basically some context for this change is useful

**Comment by narefyev91** [comment]
> https://user-images.githubusercontent.com/97473779/220686449-7b9a2941-0400-486a-90d3-b14eb556bb82.mp4
> 
> 
> 
> The issue is still reproducible in mWeb as mentioned before.
> 
> ~Do we leave this for now since mWeb keyboard is out of our control?~
> 
> I think we should fix mWeb issue here as well.

That will be pretty tricky to find a way to fix that on mWeb. We may not block this current fix for android and I will start investigating mWeb issue - it happened in different places in the app. And to find global fix will be very hard - I will do a workaround inside currency screens, and will keep posted what I will find

**Comment by mountiny** [comment]
@0xmiroslav I think we have confirmed though that behaviour is same on `main` so I am not sure if we should hold on that. We cannot fix it easily.

**Comment by mountiny** [comment]
@narefyev91 More conflicts, can you please resolve those?

**Review by 0xmiroslav — DISMISSED** [review]
This is android fix only.
Other platforms just keep same behavior as production.
Known that mWeb issues are difficult to fix, we agreed they are considered as separate issue.

File: src/components/ScreenWrapper/propTypes.js [hunk 3]
```diff
@@ -28,11 +28,15 @@ const propTypes = {
         /** Indicates when an Alert modal is about to be visible */
         willAlertModalBecomeVisible: PropTypes.bool,
     }),
+
+    /** Whether to dismiss keyboard before leave a screen */
```
Comment (mountiny) [inline src/components/ScreenWrapper/propTypes.js:32]:
```suggestion
    /** Whether to dismiss keyboard before leaving a screen */
```

**Review by mountiny — COMMENTED** [review]
One NAB

**Review by mountiny — APPROVED** [review]
Thanks!

**Pull request merged**